torch/_decomp/decompositions.py

```diff
@@ -914,7 +915,13 @@ def check_positive(param, param_name, strict=True):
 @register_decomposition(aten.native_dropout_backward)
 @pw_cast_for_opmath
 def native_dropout_backward(grad_output: Tensor, mask: Tensor, scale: float):
-    return grad_output * (mask.type_as(grad_output) * scale)
+    # According to the CUDA kernel implementation we should have this test;
+    # but it seems to fail tests!
+    # utils.check(mask.dtype == torch.bool, lambda: f"Mask should be Bool Scalar Type {mask.dtype}")
+    r = (grad_output * (mask.type_as(grad_output) * scale)).clone(
+        memory_format=utils.suggest_memory_format(grad_output)
```

Comment:
this seems like a fix unrelated to symintification right?

Reply:
yes, but it's still needed for correct stride.

Reply:
hm, this looks wrong, dropout propagates strides, so even if grad_output is not channels last, the permutation of input/grad_output would be preserved.

Reply:
For native_dropout_backward cuda kernel, the output stride doesn't follow TI's behavior. 
I suspect this is calling cudnn dropout backward. 

```
>>> torch.ops.aten.native_dropout_backward(r.permute(2,3,1,0), m, 0.2).stride()
(8, 4, 2, 1)
>>> torch.ops.aten.native_dropout_backward(r.permute(0,3,1,2), m, 0.2).stride()
(8, 1, 4, 2)
```

Reply:
For CPU kernel, it's following TI's behavior.

Reply:
CUDA's impl 

```
template <typename mask_t>
Tensor dropout_backward_cuda(const Tensor& grad, const Tensor& mask, double scale){
   Tensor ret = at::empty_like(grad, grad.suggest_memory_format());
   AT_DISPATCH_FLOATING_TYPES_AND2(at::ScalarType::Half, at::ScalarType::BFloat16, ret.scalar_type(), "masked_scale", [&] {
      using accscalar_t = acc_type<scalar_t, true>;
      masked_scale_kernel<mask_t, scalar_t>(ret, grad, mask, (accscalar_t)scale);
  });
  return ret;
}
```

Reply:
Ah I see, usually for cuda dropout `fused_dropout`/`fused_dropout_backward` (that dispatches to `masked_scale`) is used, and that one has a more reasonable stride propagation behavior.

Reply:
Hm, no, I'm confused, how does this work:
```
In [2]: a=torch.randn(1024, 1024, device="cuda").t()

In [4]: gO=torch.rand_like(out[0])

In [5]: a.requires_grad_(True)
In [6]: out=torch.native_dropout(a, p=0.5, train=True)

In [8]: out[0].backward(gO)

In [9]: a.grad.stride()
Out[9]: (1, 1024)
```
gO here can't have `suggest_memory_format` suggest anything, does permutation happen after the kernel? That's weird.